File: Intervals/Java/Merge Intervals.java
// Problem link: https://leetcode.com/problems/merge-intervals/description/
// Time: O(n log n)
// Space: O(n)
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        List<int[]> result = new ArrayList<>();
        result.add(intervals[0]);

        for(int idx = 1; idx < intervals.length; idx++) {
            int prev = result.getLast()[1];

            if(intervals[idx][0] <= prev) {
                result.getLast()[1] = Math.max(prev, intervals[idx][1]);
            }
            else {
                result.add(intervals[idx]);
            }
        }

        return result.toArray(new int[result.size()][2]);
    }
}

File: Intervals/Python/Merge Intervals.py
# Problem link: https://leetcode.com/problems/merge-intervals/description/
# Time: O(n log n)
# Space: O(n)
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        result = [intervals[0]]

        for start, end in intervals[1:]:
            prev = result[-1][1]

            if prev >= start:
                result[-1][1] = max(prev, end)
            else:
                result.append([start, end])
        
        return result
